Sources/AblyChat/Subscription.swift

```diff
@@ -4,22 +4,112 @@
 //
 // At some point we should define how this thing behaves when you iterate over it from multiple loops, or when you pass it around. I’m not yet sufficiently experienced with `AsyncSequence` to know what’s idiomatic. I tried the same thing out with `AsyncStream` (two tasks iterating over a single stream) and it appears that each element is delivered to precisely one consumer. But we can leave that for later. On a similar note consider whether it makes a difference whether this is a struct or a class.
 //
-// TODO: I wanted to implement this as a protocol (from which `MessageSubscription` would then inherit) but struggled to do so, hence the struct. Try again sometime. We can also revisit our implementation of `AsyncSequence` if we migrate to Swift 6, which adds primary types and typed errors to `AsyncSequence` and should make things easier.
-public struct Subscription<Element>: Sendable, AsyncSequence {
-    // This is a workaround for the fact that, as mentioned above, `Subscription` is a struct when I would have liked it to be a protocol. It allows people mocking our SDK to create a `Subscription` so that they can return it from their mocks. The intention of this initializer is that if you use it, then the created `Subscription` will just replay the sequence that you pass it.
-    public init<T: AsyncSequence>(mockAsyncSequence _: T) where T.Element == Element {
-        fatalError("Not implemented")
+// TODO: I wanted to implement this as a protocol (from which `MessageSubscription` would then inherit) but struggled to do so, hence the struct. Try again sometime. We can also revisit our implementation of `AsyncSequence` if we migrate to Swift 6, which adds primary types and typed errors to `AsyncSequence` and should make things easier; see https://github.com/ably-labs/ably-chat-swift/issues/21.
+public struct Subscription<Element: Sendable>: Sendable, AsyncSequence {
+    private enum Mode: Sendable {
+        case `default`(stream: AsyncStream<Element>, continuation: AsyncStream<Element>.Continuation)
+        case mockAsyncSequence(TypeErasedNonThrowingAsyncSequence)
     }
 
-    // (The below is just necessary boilerplate to get this to compile; the key point is that `next()` does not have a `throws` annotation.)
+    fileprivate struct TypeErasedNonThrowingAsyncSequence: AsyncSequence, Sendable {
+        private var makeAsyncIteratorImpl: @Sendable () -> AsyncIterator
+
+        init<T: AsyncSequence & Sendable>(asyncSequence: T) where T.Element == Element {
+            makeAsyncIteratorImpl = {
```

Comment:
Why not to keep `asyncSequence` directly and call it without this additional closure step?

Reply:
What type would you make the `asyncSequence` property in that case? It’d have to be of type `T`, meaning that `T` would have to become a generic type parameter of `TypeErasedNonThrowingAsyncSequence`, meaning that it’s no longer type-erased.

Reply:
Can't it be `AsyncSequence & Sendable` like [here](https://github.com/ably-labs/ably-chat-swift/blob/036a656653a8826b06f7f6d734d5c2cb231f67cd/Sources/AblyChat/Messages.swift#L58)?

Reply:
The code you've linked to isn’t type-safe — there are `as!` in there. This extra code is the price you pay to make it type-safe.

Reply:
(It's a good reminder that I should also implement the `MessageSubscription` type whilst I’m at it; will do that!)

Reply:
Ah okay, forgot to look two lines below, now I see 👍

Reply:
I've just tried with `var mockAsyncSequence: any AsyncSequence & Sendable` and `public init<T: AsyncSequence & Sendable>(_ mock: T)` and it works.

Reply:
Can you share the full code please? I don't see how the compiler would know the element type of `mockAsyncSequence`.

Reply:
<img width="679" alt="Screenshot 2024-08-23 at 12 08 12" src="https://github.com/user-attachments/assets/af4bff69-9a8f-4824-b2cc-3b47f5408f79">

It's in my branch where `MessageSubscription` is also `AsyncIteratorProtocol`, but it doesn't matter here.

Reply:
I’m confused. Are you referring to https://github.com/ably-labs/ably-chat-swift/blob/036a656653a8826b06f7f6d734d5c2cb231f67cd/Sources/AblyChat/Messages.swift#L55-L75? Isn't that the same code you referred to in https://github.com/ably-labs/ably-chat-swift/pull/22#discussion_r1727817815, which I said isn't type-safe?

Reply:
That one was with `as!`, the new one is without. I don't remember why I've put `as!` there, probably by mistake.

Reply:
But where can I see the full code of "the new one"?

Reply:
I didn't pushed it yet.

Reply:
This one https://github.com/ably-labs/ably-chat-swift/pull/16/commits/349a28e93ca20b802e421b9af10fc58c43c9837d

Reply:
It still has the same problem of needing to cast the element type — see `try! await mockAsyncSequence.next() as! MessageSubscription.Element`.